kernel/unwind.c

```diff
@@ -0,0 +1,582 @@
+#include "unwind.h"
+
+#include "com.h"
+#include "util.h"
+
+extern const char _eh_frame_start[];
+extern const char _eh_frame_end[];
+
+/* DWARF data structs */
+enum dwarf_rule {
+  DW_UNDEFINED,
+  DW_SAME,
+  DW_OFFSET,
+  DW_REG,
+  DW_CFA,  /* reg + offset */
+};
+
+typedef struct {
+  int32_t offset;
+  int8_t reg;
+  enum dwarf_rule rule;
+} dwarf_loc;
+
+typedef struct {
+  uint64_t loc;
+  dwarf_loc cfa;
+  dwarf_loc registers[17];
+} cfa_row;
+
+typedef struct {
+  uint64_t loc;
+  uint64_t cfa;
+  uint64_t registers[17];
+  uint8_t regs_defined[3];  /* bit array */
+} stack_frame;
+
+typedef struct {
+  uint64_t eh_data;
+  uint64_t code_af;
+  int64_t data_af;
+  uint8_t return_addr_column;
+  uint8_t aug_str_meaning;
+  uint8_t fde_encoding;
+  uint8_t lsda_encoding;
+  cfa_row init_row;
+} eh_cie;
+
+typedef struct {
+  const eh_cie *cie;
+  uint64_t pc_begin;
+  uint64_t pc_len;
+  const char *instrs;
+  uint64_t instrs_len;
+} eh_fde;
+
+/* This is a struct used all over the place in the file to parse the eh_frame
+ * section. It's a string slice with a logical head at the beginning. */
+typedef struct {
```

Comment:
All of these head functions might make more sense in a *head*er file just for them.

Also, there's a lot of repetition in the read_* methods that I wonder if they could be consolidated? This file is massive, relatively speaking.